test/common/config/registry_test.cc

```diff
@@ -182,6 +193,50 @@ TEST(RegistryTest, TestDoubleRegistrationByName) {
                             "Double registration for name: 'testing.published.test'");
 }
 
+class TestVersionedWithDeprecatedNamesFactoryAndAdditionalCategory : public PublishedFactory {
+public:
+  std::string category() const override { return "testing.published.additional.category"; }
+  std::string name() const override {
+    return "testing.published.versioned.instead_name_and_category";
+  }
+};
+
+TEST(RegistryTest, DEPRECATED_FEATURE_TEST(VersionedWithDeprecatedNamesFactoryAndNewCategory)) {
+  TestVersionedWithDeprecatedNamesFactoryAndAdditionalCategory test;
+
+  // Check the category is not registered
+  EXPECT_FALSE(Registry::FactoryCategoryRegistry::isRegistered(test.category()));
+
+  auto factory =
+      Registry::RegisterFactory<TestVersionedWithDeprecatedNamesFactoryAndAdditionalCategory,
+                                PublishedFactory>(
+          FACTORY_VERSION(0, 0, 1, {{"build.kind", "private"}}),
+          {"testing.published.versioned.deprecated_name_and_category"});
+
+  // Check the category now registered
+  EXPECT_TRUE(Registry::FactoryCategoryRegistry::isRegistered(test.category()));
+
+  const auto& factories = Envoy::Registry::FactoryCategoryRegistry::registeredFactories();
+
+  auto version =
+      factories.find("testing.published.additional.category")
+          ->second->getFactoryVersion("testing.published.versioned.instead_name_and_category");
+  EXPECT_TRUE(version.has_value());
+  EXPECT_EQ(0, version.value().version().major_number());
+  EXPECT_EQ(0, version.value().version().minor_number());
+  EXPECT_EQ(1, version.value().version().patch());
+  EXPECT_EQ(1, version.value().metadata().fields().size());
+  EXPECT_EQ("private", version.value().metadata().fields().at("build.kind").string_value());
+
+  // Get the version using deprecated name and check that it matches the
+  // version obtained through the new name.
+  auto deprecated_version =
+      factories.find("testing.published.additional.category")
+          ->second->getFactoryVersion("testing.published.versioned.deprecated_name_and_category");
+  EXPECT_TRUE(deprecated_version.has_value());
+  EXPECT_THAT(deprecated_version.value(), ProtoEq(version.value()));
```

Comment:
These would be a better candidate for using the `Optional` matcher.

Reply:
Looks like this got unexpectedly marked as resolved. I think this can be `EXPECT_THAT(deprecated_version, Optional(ProtoEq(version.value())));`

Reply:
Ooops, sorry, this was resolved in the latest commit. Thanks!